File: 213__House_Robber_II.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0
        if len(nums) <= 2:
            return max(nums)
        tmp1 = self.dp(nums[:-1])
        tmp2 = self.dp(nums[1:])
        return max(tmp1, tmp2)

    def dp(self,nums):
        dp = [0] * len(nums)
        dp[0] = nums[0]
        dp[1] = nums[1]
        for i in range(2, len(nums)):
            dp[i] = nums[i] + max(dp[:i - 1])
        return max(dp[-1], dp[-2])

File: Java/213__House_Robber_II.java
class Solution {
    public int rob(int[] nums) {
        if (nums == null || nums.length == 0)
            return 0;
        if (nums.length == 1)
            return nums[0];
        int len = nums.length;
        int[] dp = new int[len];
        dp[0] = 0;
        dp[1] = nums[0];
        for (int i = 2; i < len; i++) {
            dp[i] = Math.max(dp[i - 1], dp[i - 2] + nums[i - 1]);
        }
        int result = dp[len - 1];
        dp = new int[len];
        dp[0] = 0;
        dp[1] = nums[1];
        for (int i = 2; i < len; i++) {
            dp[i] = Math.max(dp[i - 1], dp[i - 2] + nums[i]);
        }
        result = Math.max(result, dp[len - 1]);
        return result;

    }
}
